src/inc/til/u8u16convert.h

```diff
@@ -306,25 +91,160 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+                const auto codePointLen{
+                    2 + // the code point consists of at least 2 bytes because if we cached data it will never be ASCII
+                    gsl::narrow_cast<int>((state.partials[0] & IsLead3) == IsLead3) + // add one if 3 high-order bits of the lead byte are set
+                    gsl::narrow_cast<int>((state.partials[0] & IsLead4) == IsLead4) // add another one if 4 high-order bits of the lead byte are set
+                };
+
+                const auto fetch8{ codePointLen - gsl::narrow_cast<int>(state.partials[3]) }; // number of bytes we need from the begin of the chunk to complete the code point
+                if (fetch8 > len8) // we still didn't get enough data to complete the code point, however this is not an error
+                {
+                    std::move(cursor8, cursor8 + len8, state.partials.data() + state.partials[3]);
+                    state.partials[3] += gsl::narrow_cast<char>(len8);
+                    out.clear();
+                    return S_OK;
+                }
+
+                std::move(cursor8, cursor8 + fetch8, state.partials.data() + state.partials[3]);
+                len16 = MultiByteToWideChar(CP_UTF8, 0UL, state.partials.data(), gsl::narrow_cast<int>(codePointLen), out.data(), capa16);
+                RETURN_HR_IF(E_UNEXPECTED, !len16);
+
+                state.reset();
+                capa16 -= len16;
+                len8 -= fetch8;
+                cursor8 += fetch8;
+            }
+
+            if (!!len8)
+            {
+                auto backIter{ cursor8 + len8 };
+                // If the last byte in the string was a byte belonging to a UTF-8 multi-byte character
+                if (!!(backIter[-1] & MaskAscii))
+                {
+                    // Check only up to 3 last bytes, if no Lead Byte is found then the byte before will be the Lead Byte and no partials are in the string
+                    const auto stopLen{ std::min(len8, 3) };
+                    for (auto sequenceLen{ 1 }; sequenceLen <= stopLen; ++sequenceLen)
+                    {
+                        --backIter;
+                        // If Lead Byte found
+                        if ((*backIter & MaskCont) > IsCont)
+                        {
+                            // If the Lead Byte indicates that the last bytes in the string is a partial UTF-8 code point then cache them:
+                            //  Use the bitmask at index `sequenceLen`. Compare the result with the operand having the same index. If they
+                            //  are not equal then the sequence has to be cached because it is a partial code point. Otherwise the
+                            //  sequence is a complete UTF-8 code point and the whole string is ready for the conversion into a UTF-16 string.
+                            if ((*backIter & cmpMasks[gsl::narrow_cast<size_t>(sequenceLen)]) != cmpOperands[gsl::narrow_cast<size_t>(sequenceLen)])

```

Comment:
If you moved the `codePointLen` calculation down here to store the "remaining" count, then you can drop `cmpMasks` as it shouldn't be needed anymore.